contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,964 @@
+        if (_oethbBalance > 0) {
+            IVault(vaultAddress).burnForStrategy(_oethbBalance);
+        }
+    }
+
+    /// @dev this function assumes there are no uncollected tokens in the clPool owned by the.
+    ///      strategy contract. For that reason any liquidity withdrawals also collect the tokens.
+    function _updateUnderlyingAssets() internal {
+        if (tokenId == 0) {
+            underlyingAssets = 0;
+            emit UnderlyingAssetsUpdated(underlyingAssets);
+            return;
+        }
+
+        uint128 _liquidity = _getLiquidity();
+
+        /**
+         * Our net value represent the smallest amount of tokens we are able to extract from the position
+         * given our liquidity.
+         *
+         * The least amount of tokens extraditable from the position is where the active trading price is
+         * at the ticker 0 meaning the pool is offering 1:1 trades between WETH & OETHb. At that moment the pool
+         * consists completely of OETHb and no WETH.
+         *
+         * The more swaps from WETH -> OETHb happen on the pool the more the price starts to move towards the -1
+         * ticker making OETHb (priced in WETH) more expensive.
+         */
+        (uint256 _wethAmount, uint256 _oethbAmount) = helper
+            .getAmountsForLiquidity(
+                sqrtRatioX96TickClosestToParity, // sqrtRatioX96
+                sqrtRatioX96TickLower, // sqrtRatioAX96
+                sqrtRatioX96TickHigher, // sqrtRatioBX96
+                _liquidity
+            );
+
+        require(_wethAmount == 0, "Non zero wethAmount");
+        underlyingAssets = _oethbAmount;
+        emit UnderlyingAssetsUpdated(underlyingAssets);
+    }
+
+    /**
+     * @notice Withdraw an `amount` of assets from the platform and
+     *         send to the `_recipient`.
+     * @param _recipient  Address to which the asset should be sent
+     * @param _asset      WETH address
+     * @param _amount     Amount of WETH to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external override onlyVault nonReentrant {
+        require(_asset == WETH, "Unsupported asset");
+        require(_recipient == vaultAddress, "Only withdraw to vault allowed");
+
+        uint256 _wethBalance = IERC20(WETH).balanceOf(address(this));
+        if (_wethBalance < _amount) {
+            uint256 _additionalWethRequired = _amount - _wethBalance;
+            (uint256 _wethInThePool, ) = getPositionPrincipal();
+
+            if (_wethInThePool < _additionalWethRequired) {
+                revert NotEnoughWethLiquidity(
+                    _wethInThePool,
+                    _additionalWethRequired
+                );
+            }
+
+            uint256 shareOfWethToRemove = Math.min(
+                _additionalWethRequired.divPrecisely(_wethInThePool) + 1,
+                1e18
+            );
+            _removeLiquidity(shareOfWethToRemove);
```

Comment:
🟠 `_removeLiquidity` doesn't have a `tokenId != 0` check. So, there's a chance that this call will fail. However it may happen only if Vault tries to withdraw more than what this strategy holds.

I don't see us having that problem (assuming this won't be the default strategy ever) but adding that check in `_removeLiquidity` with a proper error message would be helpful

Reply:
I think the code is correct here since `_removeLiquidity` isn't optional here but rather mandatory. It only removes liquidity when the additional token balance is required for withdrawal. 
I've added and explicit check for better error reporting: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/2981a98668f1b2a84d779fc669e410372545f00a